pyqir-parser/pyqir/parser/_parser.py

```diff
@@ -605,7 +605,7 @@ def type(self) -> QirType:
         """
         if self._type == None:
             self._type = QirType(self.instr.type)
-        return self._type
+        return cast(QirType, self._type)
```

Comment:
I wonder if we can use [https://docs.python.org/3/library/functools.html#functools.cached_property](`@functools.cached_property`) instead? That would remove the extra cache fields and the casts.

Reply:
Hmm, never mind, `cached_property` is from Python 3.8... It would still be nice to avoid `cast` in too many places since it can be used unsoundly, but that's probably out of scope for this PR.

Reply:
Another option is to refactor this to something like:

```python
if self._type is None:
    ty = QirType(self.instr.type)
    self._type = ty
else:
    ty = self._type
return ty
```

where Mypy should hopefully be able to convince itself that `ty` is not an Optional.

Reply:
Thanks @samarsha for your feedback!

Just tested the refactoring. It went through.

But I'm not a big fan of this workaround:

- The code is longer
- At first sight, it does not look efficient/straightforward
- It introduces a new unnecessary variable
- This variable itself is not type-hinted

Why do you think that cast "can be used unsoundly" ?

It is only used for type hinting.

On a personal standpoint, I'd favor the "unrefactored" version.

But I'll follow the community's conclusions, of course. 😅

Reply:
But I recognize also that the code was already using this pattern before the type hints were added, so if you don't want to refactor in this PR, I'd be OK with using `cast` for now.

Reply:
Sorry for not replying earlier.

Very instructive demonstration, thanks!

You make a point and it's a pity it adds complexity. Well, `mypy` often does...

Not having much time to correct all occurrences, I will get back to the previous version with `cast`, since it's OK for you.

It will allow us to close #132 and the specific update of removing `cast`s could be another issue/PR.

Reply:
PS : I also replaced all `== None` (including elsewhere in `_parser.py`) with `is None`.